File: src/test/java/no/nav/syfo/kafka/OversikthendelseProducerTest.java
package no.nav.syfo.kafka;

import no.nav.syfo.kafka.producer.OversikthendelseProducer;
import no.nav.syfo.kafka.producer.model.KOversikthendelse;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.util.concurrent.ListenableFuture;

import static java.time.LocalDateTime.now;
import static no.nav.syfo.kafka.producer.OversikthendelseProducer.OVERSIKTHENDELSE_TOPIC;
import static no.nav.syfo.kafka.producer.OversikthendelseType.MOTEBEHOV_SVAR_BEHANDLET;
import static no.nav.syfo.kafka.producer.OversikthendelseType.MOTEBEHOV_SVAR_MOTTATT;
import static no.nav.syfo.testhelper.UserConstants.ARBEIDSTAKER_FNR;
import static no.nav.syfo.testhelper.UserConstants.NAV_ENHET;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class OversikthendelseProducerTest {

    @Mock
    private KafkaTemplate<String, Object> kafkaTemplate;

    @InjectMocks
    private OversikthendelseProducer oversikthendelseProducer;

    @Test
    public void sendOversikthendelse() {
        when(kafkaTemplate.send(anyString(), anyString(), any(KOversikthendelse.class))).thenReturn(mock(ListenableFuture.class));

        KOversikthendelse kOversikthendelse = KOversikthendelse.builder()
                .fnr(ARBEIDSTAKER_FNR)
                .hendelseId(MOTEBEHOV_SVAR_MOTTATT.name())
                .enhetId(NAV_ENHET)
                .tidspunkt(now())
                .build();

        oversikthendelseProducer.sendOversikthendelse(kOversikthendelse);

        verify(kafkaTemplate).send(eq(OVERSIKTHENDELSE_TOPIC), anyString(), same(kOversikthendelse));
    }

    @Test
    public void sendOversikthendelseMotebehovSvarBehandlet() {
        when(kafkaTemplate.send(anyString(), anyString(), any(KOversikthendelse.class))).thenReturn(mock(ListenableFuture.class));

        KOversikthendelse kOversikthendelse = KOversikthendelse.builder()
                .fnr(ARBEIDSTAKER_FNR)
                .hendelseId(MOTEBEHOV_SVAR_BEHANDLET.name())
                .enhetId(NAV_ENHET)
                .tidspunkt(now())
                .build();

        oversikthendelseProducer.sendOversikthendelse(kOversikthendelse);

        verify(kafkaTemplate).send(eq(OVERSIKTHENDELSE_TOPIC), anyString(), same(kOversikthendelse));
    }
}

File: src/test/kotlin/no/nav/syfo/oversikthendelse/OversikthendelseProducerTest.kt
package no.nav.syfo.oversikthendelse

import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import java.util.*
import no.nav.syfo.LocalApplication
import no.nav.syfo.personoppgavehendelse.PersonoppgavehendelseProducer
import no.nav.syfo.personoppgavehendelse.domain.KPersonoppgavehendelse
import no.nav.syfo.personoppgavehendelse.domain.PersonoppgavehendelseType
import no.nav.syfo.testhelper.UserConstants.ARBEIDSTAKER_FNR
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.support.SendResult
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.util.concurrent.ListenableFuture

@ExtendWith(SpringExtension::class)
@SpringBootTest(classes = [LocalApplication::class])
@DirtiesContext
class OversikthendelseProducerTest {
    @MockkBean(relaxed = true)
    private lateinit var kafkaTemplate: KafkaTemplate<String, Any>

    @Autowired
    private lateinit var personoppgavehendelseProducer: PersonoppgavehendelseProducer

    @Test
    fun sendOversikthendelse() {
        every { kafkaTemplate.send(any(), any(), any()) } returns mockk<ListenableFuture<SendResult<String, Any>>>(
            relaxed = true
        )
        val kPersonoppgavehendelse = KPersonoppgavehendelse(
            personident = ARBEIDSTAKER_FNR,
            hendelsetype = PersonoppgavehendelseType.MOTEBEHOV_SVAR_MOTTATT.name
        )
        personoppgavehendelseProducer.sendPersonoppgavehendelse(kPersonoppgavehendelse, UUID.randomUUID())
        verify { kafkaTemplate.send(any(), any(), any()) }
    }

    @Test
    fun sendPersonoppgavehendelseMotebehovSvarBehandlet() {
        every { kafkaTemplate.send(any(), any(), any()) } returns mockk<ListenableFuture<SendResult<String, Any>>>(
            relaxed = true
        )
        val kPersonoppgavehendelse = KPersonoppgavehendelse(
            personident = ARBEIDSTAKER_FNR,
            hendelsetype = PersonoppgavehendelseType.MOTEBEHOV_SVAR_BEHANDLET.name
        )
        personoppgavehendelseProducer.sendPersonoppgavehendelse(kPersonoppgavehendelse, UUID.randomUUID())
        verify { kafkaTemplate.send(any(), any(), any()) }
    }
}
